Add glossary.md (#38)
ethereum-optimism/optimistic-specs: glossary.md

# Glossary of Optimistic Ethereum Terms

### Layer 1 (L1)

> The Ethereum Mainnet.

Reviewer: This technically could be a testnet. Perhaps a better definition could be an ethereum network that derives its security from proof of work/proof of stake

Author: ```suggestion
An Ethereum network with its own native security mechanism (ie. proof of work, proof of stake or proof of authority).
```

### Layer 2 (L2)

> The Optimistic Ethereum Network.

Reviewer: A network that derives its security from a layer one

Author: ```suggestion
A network that derives its security from a layer one.
```

> ### State Root Oracle

Reviewer: Is this supposed to be the block hash oracle? Also could link to this? https://eips.ethereum.org/EIPS/eip-210

Author: Yes, it's the same as the `Block Hash Oracle`, but we've changed the name to this (actually I think we're even shifting towards `L2 State Root Oracle`. 
Yes, it provides block hashes, but the purpose for it is more closely tied to getting state roots for L2.

Author: ```suggestion
### L2 State Root Oracle
```

> An Ethereum network with its own native security mechanism (ie. proof of work, proof of stake or proof of authority).

Reviewer 2: ```suggestion
An Ethereum network with its own native security mechanism (e.g., proof of work, proof of stake or proof of authority).
```
https://english.stackexchange.com/questions/1629/e-g-versus-i-e

A smart contract on L1, providing access to historical L1 block hashes. The existence of this contract makes it possible to prove the existence of a particular L2 state root within a previous L1 block.

### Block Generation (AKA Block Derivation)

> 

Author: Only because I was expecting this to be read during an in-person session, and that some of those people would be used to the former terminology.

All of the data necessary to reconstruct the history and state of L2 is contained within L1 blocks. Block Generation is the process of converting data contained in L1 blocks into L2 blocks.

### Epoch

A continuous sequence of 0 or more L2 blocks. There is a one to one correspondence between L2 Epochs, and L1 blocks.

> 

Author: Yes, good point!
```suggestion
A continuous sequence of L2 blocks, including a single deposit block, followed by 0 or more L2 blocks. There is a one to one correspondence between L2 Epochs, and L1 blocks.```

> 

Reviewer 3: Really not important, but is that last comma in the suggested change grammatical?

**Example usage:** "Epoch 20 corresponds to L1 block 637, epoch 21 corresponds to block 638, etc".

### Feeds

Feeds are append only lists, each of which is contained in a smart contract on Layer 1. The state of the Rollup Chain is determined by the data written to these contracts.

### Sequencer Feed

The Sequencer Feed is a contract which contains the data representing sequencer submitted blocks, each element in this feed will be used to generate a "sequencer block".

### Deposit Feed

The Deposit Feed is a contract which contains deposit data. Each element in this feed will be used to generate a "deposit block".

### Sequencing Window

The number of L1 blocks during which the Sequencer _may_ insert Sequencer blocks before a Deposit block.

If the Sequencer Window is 2 blocks, and Alice makes a Deposit in Block 420, her deposit WILL be included (by the Block Generation process) at block 422.

> 

Reviewer 3: I like this! We should probably mention the concepts of epoch: the sequencing windows is the number of L1 blocks during which the sequencer may insert L2 sequencer blocks for a given epoch, i.e. after the epoch's L2 deposit block, but before the next epoch's deposit block.

Also a nitpick, but a sequencing window is a block range whose *size* is the number of L1 blocks [...]. e.g. the epoch for L1 block w/ number `X` is `[X+1, X+k]` where `k` is the window size.

### Execution Environment (EE)

> 

Reviewer 3: Isn't EE "Execution Engine"?

The virtual machine in which transactions are applied to the pre-state to generate the post-state.

### Rollup Node AKA Rollup Driver

> 

Reviewer 3: We should probably choose between both of these - no particular opinion here.

The core client software which reads data from an L1 node, and transforms it into L2 data. Both the Sequencer, and Verifier nodes run a Rollup Node, albeit in different configurations.

> 

Reviewer 3: As far as I understand (notably referring to this diagram we charted together), the tx batch submitted and proposal submitter are also part of the rollup node, and go on the opposite direction from this (they submit L2 tx and block hashes to L1).

We might want to be more precise thant just "data" here. In particular, the thing being read from L1 are "block inputs" (or some other term we come to agreement on, but the goal is to do block derivation).

### Consensus Layer

By analogy to the architecture of ETH2 we can also describe Layer 1 as the "Consensus Layer", as it defines the canonical history and state of the Layer 2 chain.

> 

Reviewer 3: Do we want to add these analogies / overloaded terms? I think if we do this it becomes confusing because suddenly "consensus layer" can mean either the L1 consensus layer or the whole of L1.

### Sequencer

The privileged entity which accepts L2 transactions from users, and determines the ordering of those transactions with respect to each other and deposit transactions.

### Verifier

A verifier is any Rollup Node operator who is not the Sequencer. Verifiers compute the state of the L2 chain, and compare it to the states proposed by the Sequencer.

> 

Reviewer 3: Could be good to reuse part of the (existing) intro here that lists the goals of the verifier (validating + fraud proof, distributing state to users, and amongst validators).

### Deposit transaction (via the Deposit Feed)

A transaction which originates on L1, by way of a message sent to the Deposit Feed contract. This message can be sent either by a contract or an externally owned account.

### Sequencer transaction (via the Sequencer feed)

> A transaction which is sent directly to the sequencer from user.

Reviewer 3: user**s**

### L2 State Oracle (AKA L2 Block Oracle)

> 

Reviewer 3: L2 proposal manager (cf. our conversation about how I didn't like "oracle" because it implies a very simple question-answer interface which is not the case here)
